### ISSUE TITLE
Defining a custom loadBalancerSourceRanges in a AWS NLB service is not respected

### ISSUE THREAD

**Issue by aledbf** [kubernetes/kubernetes#57212]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Creating a NLB service with a custom `loadBalancerSourceRanges` is not respected

**What you expected to happen**:

The security group should be created with the defined range and not `0.0.0.0/0`

**Comment by aledbf** [comment]
ping @micahhausler

**Comment by 2rs2ts** [comment]
NLBs don't have security groups

**Comment by 2rs2ts** [comment]
If you wanted to accomplish this with the NLB k8s would have to manipulate the security groups of the nodes themselves.

**Comment by micahhausler** [comment]
That is what the controller does right now

**Comment by 2rs2ts** [comment]
Do you mean that's how it handles the classic ELB LoadBalancer Service types? Because it changes the rules on the SGs on the ELB for that. Or do you mean something else? (Forgive my ignorance, I'm merely trying to be helpful)

**Comment by micahhausler** [comment]
You are right, NLBs do not have Security Groups. The current NLB controller opens up the nodePort on the nodes' security group to `0.0.0.0/0` in order to allow traffic. When a user defines `loadBalancerSourceRanges`, it should respect the IP CIDRs they specify.

Classic ELBs do have security groups, that is not covered in this issue.

**Comment by 2rs2ts** [comment]
gotcha, thanks, sorry for the distraction

**Comment by jrnt30** [comment]
tl;dr - The current controller doesn't seem to reconcile updates to the `loadBalancerSourceRanges` for existing target groups properly however it *does* seem to do the right thing if there is a new target group created in conjunction with a change.

## Initial Deployment
Our initial deployment did not include any restrictions via `loadBalancerSourceRanges`.

```apiVersion: v1
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "60"
    service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true"
    service.beta.kubernetes.io/aws-load-balancer-type: nlb
  name: nginx-ingress-controller
  namespace: kube-system
spec:
  externalTrafficPolicy: Local
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: http
  - name: https
    port: 443
    protocol: TCP
    targetPort: https
  selector:
    app: nginx-ingress
    component: controller
    release: nginx-ingress
  sessionAffinity: None
  type: LoadBalancer
```

**Resultant Security Group Ingress Rules**
- `healthCheckNodePort` created and restricted to VPC CIDR
- `nodePort`s for `http` and `https` created and unrestricted (`0.0.0.0/0`)

## Update - Add Net New `loadBalancerSourceRanges`
We had a desire then to lock down our ingress and added in the `loadBalancerSourceRanges` and simply added the `loadBalancerSourceRanges` to the `spec`.  This did *not* have any changes on the Security Group for the nodes at all.

**Resultant Security Group Ingress Rules**
- `healthCheckNodePort` created and restricted to VPC CIDR
- `nodePort` for `http` and `https` created and unrestricted (`0.0.0.0/0`)

## Forced Update - Adjust `ports` by removing `nodePort`
We tried "tricking" the Controller into doing additional work by removing the `nodePort` attribute.  This *does* result in a meaningful change being propagated to the Security Group *however it also* means that a new Target Group is created which requires the heath checks to initialize and results in traffic being dropped until that finishes.

**Resultant Security Group Ingress Rules**
- `healthCheckNodePort` created and restricted to VPC CIDR
- New `nodePort` for modified has explicit ingress rules for IPs listed in `loadBalancerSourceRanges`
- Original `nodePort`s * still have* unrestricted `0.0.0.0/0` rule present

## Update - Append new IP to `loadBalancerSourceRanges`
Adding in another IP to the list did *not* result in any changes 

**Resultant Security Group Ingress Rules**
- `healthCheckNodePort` created and restricted to VPC CIDR
- `nodePort` has ingress for *only* the initial IP that was defined in the previous step.  
- *New IP* was *not*  added to the Security Group ingress rules

## Update - Remove all IPs 
Finally we tried just deleting the `loadBalancerSourceRanges` all together.  This resulted in no changes to the rules. 

**Resultant Security Group Ingress Rules**
- `healthCheckNodePort` created and restricted to VPC CIDR
- `nodePort` has ingress for *only* the initial IP that was defined in the previous step.  
- *No rule created for 0.0.0.0/0*

**Comment by M00nF1sh** [comment]
I'm working a fix on this, should be available nearly next week.

**Comment by tewing-riffyn** [comment]
Hi @M00nF1sh, Thank you for the PR.  Do you know when it's expected to be merged and available to test?

I'm anxiously awaiting this merge so we can use elasticIP's to offer static addresses to our web clients.  Currently we use ELBs along with a pod annotation to assign an additional handcrafted security group containing our whitelist.  The update in this ticket will allow us to trust that loadBalancerSourceRanges in our helm values are propagated to the node SG.

I'm wondering what happens if 5 NLBs point to the same worker nodes, each using 2 ports and each NLB service contains 15 loadBalancerSourceRanges.  This would make 5 x 2 x 15 = 150 rules in a single SG.  A single SG can't contain them all and I assume it will fail to add the SG rules beyond the maximum.    I'm curious if the failure would be silent or if the failure to add all SG rules would cause the service to fail creation.

AWS documentation says I can get the number of rules per SG increased (default 60) , and I can increase or decrease the maximum number of SGs allowed to be attached to a network interface but 
the Security groups per network interface limit multiplied by the Rules per security group limit can't exceed the hard limit of 1000.
https://aws.amazon.com/premiumsupport/knowledge-center/increase-security-group-rule-limit/

This limit applies across the entire AWS account so I need to consider other non-k8s deployments in our AWS account that rely on using 5 attached SGs.  This tells me I can ask AWS to increase the number of rules-per-SG to 200 without decreasing from 5 SGs per interface we use elsewhere.

Thank you,
-Terry

**Comment by tewing-riffyn** [comment]
@jrnt30 - Your Jan 30 post in this issue should win an award.

**Comment by joshimhoff** [comment]
+1 to @tewing-riffyn's question about when @M00nF1sh's PR will be ready for use in EKS.
